File: internal/application/dto/entity/recipe_ingredient_dto.go
package entity

import "github.com/vini/costify-go/internal/domain/valueobject"

// RecipeIngredientDto represents the data transfer object for a recipe ingredient
type RecipeIngredientDto struct {
	IngredientId string  `json:"ingredientId"`
	Quantity     float64 `json:"quantity"`
	Unit         string  `json:"unit"`
}

// FromRecipeIngredient creates a RecipeIngredientDto from a domain value object
func FromRecipeIngredient(ingredient valueobject.RecipeIngredient) RecipeIngredientDto {
	return RecipeIngredientDto{
		IngredientId: ingredient.IngredientId().Value(),
		Quantity:     ingredient.Quantity(),
		Unit:         ingredient.Unit().String(),
	}
}

// FromRecipeIngredients creates a slice of RecipeIngredientDto from domain value objects
func FromRecipeIngredients(ingredients []valueobject.RecipeIngredient) []RecipeIngredientDto {
	dtos := make([]RecipeIngredientDto, len(ingredients))
	for i, ingredient := range ingredients {
		dtos[i] = FromRecipeIngredient(ingredient)
	}
	return dtos
}

File: src/main/java/br/unifor/costify/application/dto/response/RecipeIngredientDto.java
package br.unifor.costify.application.dto.response;

import br.unifor.costify.domain.valueobject.RecipeIngredient;
import br.unifor.costify.domain.valueobject.Unit;

/**
 * DTO for recipe ingredient information in API responses.
 * Converts domain RecipeIngredient to a JSON-friendly format with ingredientId as a string.
 */
public record RecipeIngredientDto(
        String ingredientId,
        double quantity,
        Unit unit
) {
    public static RecipeIngredientDto from(RecipeIngredient recipeIngredient) {
        return new RecipeIngredientDto(
                recipeIngredient.getIngredientId().getValue(),
                recipeIngredient.getQuantity(),
                recipeIngredient.getUnit()
        );
    }
}
